File: algorithm/leetcodeObjectiveC/merge-Intervals.m
// 56. Merge Intervals
// Difficulty: Hard

// Given a collection of intervals, merge all overlapping intervals.

// For example,
// Given [1,3],[2,6],[8,10],[15,18],
// return [1,6],[8,10],[15,18].

// Time:  O(nlogn)
// Space: O(1)

#import <Foundation/Foundation.h>

@interface Interval : NSObject

@property (assign) NSInteger start;
@property (assign) NSInteger end;

-(id) initWithStart:(NSInteger)start end:(NSInteger)end;
-(id) init;

@end

@implementation Interval

-(id) initWithStart:(NSInteger)start end:(NSInteger)end {
  self = [super init];
  if (self) {
    self.start = start;
    self.end = end;
  }
  return self;
}

-(id) init {
  return [self initWithStart:0 end:0];
}

@end

NSArray* merge(NSArray* intervals) {
  if ([intervals count] <= 1) {
    return intervals;
  }
  
  NSArray* sorted = [intervals sortedArrayUsingComparator:^NSComparisonResult(id  _Nonnull obj1, id  _Nonnull obj2) {
    NSNumber* start1 = [NSNumber numberWithInteger: [(Interval*)obj1 start]];
    NSNumber* start2 = [NSNumber numberWithInteger: [(Interval*)obj2 start]];
    return [start1 compare: start2];
  }];
  
  NSMutableArray* result = [NSMutableArray arrayWithObject:sorted[0]];
  for (int i = 1; i < [sorted count]; ++i) {
    if ([(Interval*)[sorted objectAtIndex:i] start] <= [(Interval*)[result lastObject] end]) {
      [[result lastObject] setEnd:MAX([(Interval*)[result lastObject] end], [(Interval*)[sorted objectAtIndex:i] end])];
    } else {
      [result addObject:sorted[i]];
    }
  }
  
  return result;
}

File: zZ/java/merge-intervals.java
/**
 * Merge Intervals
 * 
 * Given a collection of intervals, merge all overlapping intervals.
 * 
 * For example,
 * Given [1,3],[2,6],[8,10],[15,18],
 * return [1,6],[8,10],[15,18].
 */

public class Solution {
    private static class MyComparator implements Comparator<Interval> {
        @Override
        public int compare(Interval interval1, Interval interval2) {
            return interval1.start - interval2.start;
        }
    }

    public List<Intervval> merge(List<Interval> intervals) {
        if (intervals == null) {
            throw new NullPointerException();
        }

        List<Interval> result = new ArrayList<Interval>();
        Collections.sort(intervals, new MyComparator());

        Iterator<Interval> iterator = intervals.iterator();
        Interval prev;
        if (iterator.hasNext()) {
            prev = iterator.next();
            result.add(prev);
        }

        while (iterator.hasNext()) {
            Interval curr = iterator.next(); 
            if (prev.end < curr.start) {
                result.add(curr);
                prev = curr;
            } else {
                prev.end = Math.max(prev.end, curr.end);
            }
        }
        
        return result;
    }
}

public class Interval {
    int start;
    int end;
    Interval() {
        start = 0;
        end = 0;
    }
    Interval(int start, int end) {
        this.start = start;
        this.end = end;
    }
}